include/node.tcc

```diff
@@ -108,71 +192,61 @@ template <unsigned Tdim, unsigned Tdof, unsigned Tnphases>
 void mpm::Node<Tdim, Tdof, Tnphases>::update_external_force(
     bool update, unsigned phase,
     const Eigen::Matrix<double, Tdim, 1>& force) noexcept {
-  // Assert
-  assert(phase < Tnphases);
-
-  // Decide to update or assign
-  const double factor = (update == true) ? 1. : 0.;
-
-  // Update/assign external force
-  std::lock_guard<std::mutex> guard(node_mutex_);
-  external_force_.col(phase) = external_force_.col(phase) * factor + force;
+  this->update_vector_property(mpm::properties::Vector::ExternalForce, update,
+                               phase, force);
 }
 
 //! Update internal force (body force / traction force)
 template <unsigned Tdim, unsigned Tdof, unsigned Tnphases>
 void mpm::Node<Tdim, Tdof, Tnphases>::update_internal_force(
     bool update, unsigned phase,
     const Eigen::Matrix<double, Tdim, 1>& force) noexcept {
-  // Assert
-  assert(phase < Tnphases);
-
-  // Decide to update or assign
-  const double factor = (update == true) ? 1. : 0.;
-
-  // Update/assign internal force
-  std::lock_guard<std::mutex> guard(node_mutex_);
-  internal_force_.col(phase) = internal_force_.col(phase) * factor + force;
+  this->update_vector_property(mpm::properties::Vector::InternalForce, update,
+                               phase, force);
 }
 
 //! Assign nodal momentum
 template <unsigned Tdim, unsigned Tdof, unsigned Tnphases>
 void mpm::Node<Tdim, Tdof, Tnphases>::update_momentum(
     bool update, unsigned phase,
     const Eigen::Matrix<double, Tdim, 1>& momentum) noexcept {
-  // Assert
-  assert(phase < Tnphases);
-
-  // Decide to update or assign
-  const double factor = (update == true) ? 1. : 0.;
-
-  // Update/assign momentum
-  std::lock_guard<std::mutex> guard(node_mutex_);
-  momentum_.col(phase) = momentum_.col(phase) * factor + momentum;
+  this->update_vector_property(mpm::properties::Vector::Momentum, update, phase,
+                               momentum);
 }
 
 //! Update pressure at the nodes from particle
 template <unsigned Tdim, unsigned Tdof, unsigned Tnphases>
 void mpm::Node<Tdim, Tdof, Tnphases>::update_mass_pressure(
-    unsigned phase, double mass_pressure) noexcept {
-  // Assert
-  assert(phase < Tnphases);
+    bool update, unsigned phase, double mass_pressure) noexcept {
+  this->update_scalar_property(mpm::properties::Scalar::MassPressure, update,
+                               phase, mass_pressure);
+}
 
+//! Compute pressure from mass pressure
+//! pressure = mass pressure / mass
+template <unsigned Tdim, unsigned Tdof, unsigned Tnphases>
+void mpm::Node<Tdim, Tdof, Tnphases>::compute_pressure() {
   const double tolerance = 1.E-16;
-  // Compute pressure from mass*pressure
-  if (mass_(phase) > tolerance) {
-    std::lock_guard<std::mutex> guard(node_mutex_);
-    pressure_(phase) += mass_pressure / mass_(phase);
+  for (unsigned phase = 0; phase < Tnphases; ++phase) {
+    if (this->mass(phase) > tolerance) {
+      scalar_properties_.at(mpm::properties::Scalar::Pressure)(phase) =
```

Comment:
Don't we have pressure as a state variable?

Reply:
Yes, that's on `particle`, but on `node` we store it as `mpm::properties`.